### PR TITLE
feature/user-oauth/refactor signout controller

### PR DESCRIPTION
**Pull request by ICraig-CH**
__Short description outlining key changes/additions__

- Added two methods handling the signout process in the coordinator.

- Added tests for the methods.

- Refactored the SignoutController to use the coordinator class.

This branch was created using git cherry-pick due to merge conflicts that would arise because the original branch was not branched from feature/user-oauth/ci but instead from one that would later be merged in. 

__JIRA Ticket Number__
FA - 755 Refactor Signout Controller
### Type of change

* [ ] Bug fix
* [x] New feature
* [x] Breaking change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

File: src/main/java/uk/gov/ch/developer/docs/controller/developer/SignOutController.java [hunk 1]
```diff
@@ -32,8 +30,9 @@ public void doSignOut(final HttpServletResponse httpServletResponse,
 
         final Session chSession = (Session) httpServletRequest
                 .getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
-        oauth2.invalidateSession(chSession);
-        httpServletResponse.sendRedirect(identityProviders.getRedirectUriPage());
+        String redirectUri = ioAuthCoordinator.getSignoutUri();
+        ioAuthCoordinator.invalidateSession(chSession);
+        httpServletResponse.sendRedirect(redirectUri);
```
Comment (CH-JonathanBlake) [inline src/main/java/uk/gov/ch/developer/docs/controller/developer/SignOutController.java:35]: I'd rather that we just use the "redirect:"+ redirectUri
It removes the IOException from this class and allows Spring to handle it.

File: src/main/java/uk/gov/ch/oauth/IOAuthCoordinator.java [hunk 2]
```diff
@@ -1,11 +1,25 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.util.Map;
 
 public interface IOAuthCoordinator {
 
     String getPostCallbackRedirectURL(HttpServletResponse response, Map<String, String> params)
             throws UnauthorisedException;
+
+    /**
+     * @return The redirect Uri so that it can be assigned to a parameter for the sendRedirect() method
```
Comment (CH-JonathanBlake) [inline src/main/java/uk/gov/ch/oauth/IOAuthCoordinator.java:15]: May as well add a description. However I shouldn't have forgotten to do this for Callback. Could I ask you to cover that as well?

Comment (ICraig-CH) [reply src/main/java/uk/gov/ch/oauth/IOAuthCoordinator.java:26]: Ive added that in now :)

File: src/main/java/uk/gov/ch/developer/docs/controller/developer/SignOutController.java (hunk 1 above)
Comment (ICraig-CH) [reply src/main/java/uk/gov/ch/developer/docs/controller/developer/SignOutController.java:35]: Do you mean have the controller method return a String with that in it?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 3]
```diff
@@ -42,6 +23,15 @@
...
 
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.*;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.mockito.Mockito.*;
+
```
Comment (ch-rfishwick) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: These changes seem to be causing merge conflicts. I'm not quite why there have been import changes on this file, maybe a merge went wrong somewhere?

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: Im not sure, I tried changing and committing but it wouldn't let me

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: Should I try and resolve it on the web-editor below?

Comment (ch-rfishwick) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: Yeah, could be the way to go. What's happening when you try to make the changes on your local editor?

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: I made the changes and added a space somewhere and it let me push but for some reason it did not change the imports. Mabye Intellij "cleans" them if it notices multiple imports from the same package to a *

Comment (ch-rfishwick) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:34]: There might be something with you commit settings. Mine is set to optimise imports and cleanup the code. Are you able to see what your settings are?

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:38]: Ive unchecked optimise imports and its cleared some of them

Comment (ch-rfishwick) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:38]: I you manually revert the imports to how they supposedly were before all your changes and then commit and push again without optimise imports.....or is that what you've just done?

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:38]: I then committed with each one imported, I think the code conventions said don't use .* in imports and then committed and it cleared it. There is another part that is conflicting which is not on my local version. I think that may be the new stuff Zain merged in to /ci

Comment (ch-rfishwick) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:39]: Have you merged Zain's most recent changes into your current branch?

Comment (ICraig-CH) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java]: That worked :)

**Pull request merged**